File: Javascript contract deployment env/scripts/interaction.js
require('dotenv').config();
const { ethers } = require('hardhat');
const fs = require('fs');
const path = require('path');

async function main() {
    const [deployer] = await ethers.getSigners();
    console.log('Interacting with contracts using the account:', deployer.address);

    // Load deployed addresses
    const addresses = JSON.parse(fs.readFileSync('deployedAddresses.json'));

    const CalFundToken = await ethers.getContractFactory('CalFundToken');
    const cal = CalFundToken.attach(addresses.cal);

    const wethArtifactPath = path.join(__dirname, '../contracts/WETH9.json');
    const routerArtifactPath = path.join(__dirname, '../contracts/UniswapV2Router02.json');
    const pairArtifactPath = path.join(__dirname, '../contracts/UniswapV2Pair.json');

    const wethArtifact = JSON.parse(fs.readFileSync(wethArtifactPath));
    const routerArtifact = JSON.parse(fs.readFileSync(routerArtifactPath));
    const pairArtifact = JSON.parse(fs.readFileSync(pairArtifactPath));

    const weth = new ethers.Contract(addresses.weth, wethArtifact.abi, deployer);
    const router = new ethers.Contract(addresses.router, routerArtifact.abi, deployer);
    const pair = new ethers.Contract(addresses.pair, pairArtifact.abi, deployer);

    // Mint WETH
    const mintWethTx = await weth.deposit({ value: ethers.utils.parseUnits('2', 18) });
    await mintWethTx.wait();
    console.log('Minted 2 WETH');

    // Display deployer's balance
    const calBalance = await cal.balanceOf(deployer.address);
    const wethBalance = await weth.balanceOf(deployer.address);
    console.log('Deployer CAL balance:', ethers.utils.formatUnits(calBalance, 18));
    console.log('Deployer WETH balance:', ethers.utils.formatUnits(wethBalance, 18));

    // Add liquidity
    const calAmount = ethers.utils.parseUnits('7500', 18); // 7500 CAL
    const wethAmount = ethers.utils.parseUnits('2', 18); // 2 WETH
    const deadline = Math.floor(Date.now() / 1000) + (10 * 60); // 10 minutes from the current Unix time

    // Approve tokens for liquidity addition
    await cal.approve(router.address, calAmount).then(tx => tx.wait());
    await weth.approve(router.address, wethAmount).then(tx => tx.wait());

    const addLiquidityTx = await router.addLiquidity(
        cal.address,
        weth.address,
        calAmount,
        wethAmount,
        0,
        0,
        deployer.address,
        deadline,
        { gasLimit: ethers.utils.hexlify(1000000) }
    );
    await addLiquidityTx.wait();
    console.log('Liquidity added');

    // Fetch and display reserves before the swap
    let reserves = await pair.getReserves();
    console.log('Reserves before swap:', reserves);
    console.log('Pair CAL balance before swap:', ethers.utils.formatUnits(reserves._reserve1, 18));
    console.log('Pair WETH balance before swap:', ethers.utils.formatUnits(reserves._reserve0, 18));

    // Perform a swap from CAL to WETH
    const amountInCal = ethers.utils.parseUnits('500', 18); // 500 CAL
    const amountOutMinWeth = 0; // Accept any amount of WETH

    await cal.approve(router.address, amountInCal).then(tx => tx.wait());
    const swapPathCalToWeth = [cal.address, weth.address];
    const calToWethSwapTx = await router.swapExactTokensForTokens(
        amountInCal,
        amountOutMinWeth,
        swapPathCalToWeth,
        deployer.address,
        deadline,
        { gasLimit: ethers.utils.hexlify(1000000) }
    );
    await calToWethSwapTx.wait();
    console.log('CAL to WETH swap executed');

    // Fetch and display reserves after the first swap
    reserves = await pair.getReserves();
    console.log('Reserves after CAL to WETH swap:', reserves);
    console.log('Pair CAL balance after CAL to WETH swap:', ethers.utils.formatUnits(reserves._reserve1, 18));
    console.log('Pair WETH balance after CAL to WETH swap:', ethers.utils.formatUnits(reserves._reserve0, 18));

    // Perform a swap from WETH to CAL
    const amountInWeth = ethers.utils.parseUnits('0.2', 18); // 0.2 WETH
    const amountOutMinCal = 0; // Accept any amount of CAL

    await weth.approve(router.address, amountInWeth).then(tx => tx.wait());
    const swapPathWethToCal = [weth.address, cal.address];
    const wethToCalSwapTx = await router.swapExactTokensForTokens(
        amountInWeth,
        amountOutMinCal,
        swapPathWethToCal,
        deployer.address,
        deadline,
        { gasLimit: ethers.utils.hexlify(1000000) }
    );
    await wethToCalSwapTx.wait();
    console.log('WETH to CAL swap executed');

    // Fetch and display reserves after the second swap
    reserves = await pair.getReserves();
    console.log('Reserves after WETH to CAL swap:', reserves);
    console.log('Pair CAL balance after WETH to CAL swap:', ethers.utils.formatUnits(reserves._reserve1, 18));
    console.log('Pair WETH balance after WETH to CAL swap:', ethers.utils.formatUnits(reserves._reserve0, 18));

    // Display deployer's new balance
    const newCalBalance = await cal.balanceOf(deployer.address);
    const newWethBalance = await weth.balanceOf(deployer.address);
    console.log('New CAL balance:', ethers.utils.formatUnits(newCalBalance, 18));
    console.log('New WETH balance:', ethers.utils.formatUnits(newWethBalance, 18));
}

main()
    .then(() => process.exit(0))
    .catch((error) => {
        console.error(error);
        process.exit(1);
    });

File: EVM Blockchain/scripts/interaction.py
from brownie import accounts, network, web3, CalFundToken, interface
import json
import os
from dotenv import load_dotenv
from pathlib import Path

def main():
    # Load environment variables
    load_dotenv()
    gas_price = 2 * 10**9  # 2 gwei in wei
    gas_limit = 3000000  # Adjust as necessary
    # Set up the deployer account
    deployer = accounts.load('test2')
    print('Interacting with contracts using the account:', deployer.address)

    # Load the CalFundToken address from mintaddy.txt
    with open('mintaddy.txt', 'r') as f:
        cal_address = f.read().strip()

    # Load deployed addresses
    with open('deployedAddresses.json') as f:
        addresses = json.load(f)

    # Load the pair address from coinpair.txt
    with open('coinpair.txt', 'r') as f:
        pair_address = f.read().strip()

    # Set up the contract objects
    cal = CalFundToken.at(cal_address)
    router = interface.IUniswapV2Router02(addresses['router'])
    pair = interface.IUniswapV2Pair(pair_address)
    weth = interface.IWETH(addresses['weth'])

    # Mint WETH
    # mint_weth_tx = weth.deposit({'from': deployer, 'value': web3.to_wei(2, 'ether')})
    # mint_weth_tx.wait(1)
    # print('Minted 2 WETH')

    # Display deployer's balance
    cal_balance = cal.balanceOf(deployer.address)
    weth_balance = weth.balanceOf(deployer.address)
    print('Deployer CAL balance:', web3.from_wei(cal_balance, 'ether'))
    print('Deployer WETH balance:', web3.from_wei(weth_balance, 'ether'))

    # Add liquidity (commented out)
    # cal_amount = web3.to_wei(7500, 'ether')  # 7500 CAL
    # weth_amount = web3.to_wei(2, 'ether')  # 2 WETH
    # deadline = web3.eth.get_block('latest').timestamp + (10 * 60)  # 10 minutes from the current Unix time

    # Approve tokens for liquidity addition (commented out)
    # cal.approve(router.address, cal_amount, {'from': deployer}).wait(1)
    # weth.approve(router.address, weth_amount, {'from': deployer}).wait(1)

    # add_liquidity_tx = router.addLiquidity(
    #     cal.address,
    #     weth.address,
    #     cal_amount,
    #     weth_amount,
    #     0,
    #     0,
    #     deployer.address,
    #     deadline,
    #     {'from': deployer, 'gas_limit': 1000000}
    # )
    # add_liquidity_tx.wait(1)
    # print('Liquidity added')

    # Fetch and display reserves before the swap
    reserves = pair.getReserves()
    print('Reserves before swap:', reserves)
    print('Pair CAL balance before swap:', web3.from_wei(reserves[0], 'ether'))
    print('Pair WETH balance before swap:', web3.from_wei(reserves[1], 'ether'))

    # Perform a swap from CAL to WETH
    amount_in_cal = web3.to_wei(500, 'ether')  # 500 CAL
    amount_out_min_weth = 0  # Accept any amount of WETH

    cal.approve(router.address, amount_in_cal, {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}).wait(1)
    swap_path_cal_to_weth = [cal.address, weth.address]
    cal_to_weth_swap_tx = router.swapExactTokensForTokens(
        amount_in_cal,
        amount_out_min_weth,
        swap_path_cal_to_weth,
        deployer.address,
        (web3.eth.get_block('latest').timestamp + (10 * 60)),
        {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}
    )
    cal_to_weth_swap_tx.wait(1)
    print('CAL to WETH swap executed')

    # Fetch and display reserves after the first swap
    reserves = pair.getReserves()
    print('Reserves after CAL to WETH swap:', reserves)
    print('Pair CAL balance after CAL to WETH swap:', web3.from_wei(reserves[0], 'ether'))
    print('Pair WETH balance after CAL to WETH swap:', web3.from_wei(reserves[1], 'ether'))

    # Perform a swap from WETH to CAL
    amount_in_weth = web3.to_wei(0.1, 'ether')  # 0.2 WETH
    amount_out_min_cal = 0  # Accept any amount of CAL

    weth.approve(router.address, amount_in_weth, {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}).wait(1)
    swap_path_weth_to_cal = [weth.address, cal.address]
    weth_to_cal_swap_tx = router.swapExactTokensForTokens(
        amount_in_weth,
        amount_out_min_cal,
        swap_path_weth_to_cal,
        deployer.address,
        (web3.eth.get_block('latest').timestamp + (10 * 60)),
        {'from': deployer, 'gas_price': gas_price, 'gas_limit': gas_limit}
    )
    weth_to_cal_swap_tx.wait(1)
    print('WETH to CAL swap executed')

    # Fetch and display reserves after the second swap
    reserves = pair.getReserves()
    print('Reserves after WETH to CAL swap:', reserves)
    print('Pair CAL balance after WETH to CAL swap:', web3.from_wei(reserves[0], 'ether'))
    print('Pair WETH balance after WETH to CAL swap:', web3.from_wei(reserves[1], 'ether'))

    # Display deployer's new balance
    new_cal_balance = cal.balanceOf(deployer.address)
    new_weth_balance = weth.balanceOf(deployer.address)
    print('New CAL balance:', web3.from_wei(new_cal_balance, 'ether'))
    print('New WETH balance:', web3.from_wei(new_weth_balance, 'ether'))

def load_dotenv():
    from dotenv import load_dotenv
    load_dotenv()
